File: facecat_cs/service/FCHttpMonitor.cs
/*捂脸猫FaceCat框架 v1.0
 1.创始人-矿洞程序员-上海宁米科技创始人-脉脉KOL-陶德 (微信号:suade1984);
 2.联合创始人-上海宁米科技创始人-袁立涛(微信号:wx627378127);
 3.联合创始人-河北思尔企业管理咨询有限公司合伙人-肖添龙(微信号:xiaotianlong_luu);
 4.联合开发者-陈晓阳(微信号:chenxiaoyangzxy)，助理-朱炜(微信号:cnnic_zhu);
 5.该框架开源协议为BSD，欢迎对我们的创业活动进行各种支持，欢迎更多开发者加入。
 包含C/C++,Java,C#,iOS,MacOS,Linux六个版本的图形和通讯服务框架。
 */

using System;
using System.Collections.Generic;
using System.Text;
using System.Net;
using System.IO;
using System.Web;
using System.Threading;
using OwLib;
using System.Net.Sockets;
using FaceCat;

namespace FaceCat {
    /// <summary>
    /// Http数据
    /// </summary>
    public class FCHttpData {
        public bool m_close;
        public int m_contentLength;
        public String m_contentType = "";
        public byte[] m_body;
        public String m_method = "";
        public HashMap<string, string> m_parameters = new HashMap<string, string>();
        public byte[] m_resBytes;
        public String m_resStr;
        public String m_remoteIP;
        public int m_remotePort;
        public int m_socketID;
        public int m_statusCode = 200;
        public String m_url = "";
    }

    /// <summary>
    /// 文件Http服务
    /// </summary>
    public class HttpFileData {
        /// <summary>
        /// 流
        /// </summary>
        public byte[] m_buffer;

        /// <summary>
        /// 已收到长度
        /// </summary>
        public long m_recvLength;

        /// <summary>
        /// 总长度
        /// </summary>
        public long m_totalLength;
    }

    /// <summary>
    /// Node服务
    /// </summary>
    public class FCHttpMonitor {
        /// <summary>
        /// 创建Node服务
        /// </summary>
        /// <param name="fileName">文件名</param>
        public FCHttpMonitor(string fileName) {
            m_fileName = fileName;
        }

        /// <summary>
        /// 简单服务
        /// </summary>
        public static HashMap<String, FCHttpEasyService> m_easyServices = new HashMap<String, FCHttpEasyService>();

        /// <summary>
        /// 文件名称
        /// </summary>
        private String m_fileName;

        /// <summary>
        /// 脚本集合
        /// </summary>
        public Stack<FCScript> m_indicators = new Stack<FCScript>();

        /// <summary>
        /// 数据集合
        /// </summary>
        public HashMap<int, FCHttpData> m_httpDatas = new HashMap<int, FCHttpData>();

        private FCScript m_indicator;

        /// <summary>
        /// 获取主脚本
        /// </summary>
        public FCScript Indicator {
            get { return m_indicator; }
        }

        private Socket m_listener;

        /// <summary>
        /// 获取监听对象
        /// </summary>
        public Socket Listener {
            get { return m_listener; }
        }

        private static FCHttpMonitor m_mainMonitor;

        /// <summary>
        /// 获取主监视器
        /// </summary>
        /// <returns></returns>
        public static FCHttpMonitor MainMonitor {
            get { return m_mainMonitor; }
        }

        private int m_port = 8080;

        /// <summary>
        /// 获取或设置端口
        /// </summary>
        public int Port {
            get { return m_port; }
            set { m_port = value; }
        }

        private string m_script;

        /// <summary>
        /// 获取脚本
        /// </summary>
        public string Script {
            get { return m_script; }
        }

        private FCNative m_native;

        /// <summary>
        /// 获取XML对象
        /// </summary>
        public FCNative Native {
            get { return m_native; }
        }

        private bool m_useScript;

        /// <summary>
        /// 是否使用脚本
        /// </summary>
        public bool UseScript {
            get { return m_useScript; }
        }

        /// <summary>
        /// 检查脚本
        /// </summary>
        public void checkScript() {
            String newScript = "";
            FCFile.read(m_fileName, ref newScript);
            if (m_script != newScript) {
                Console.WriteLine("检测到脚本被修改...");
                m_script = newScript;
                lock (m_indicators) {
                    while (true) {
                        try {
                            FCScript indicator = m_indicators.Pop();
                            indicator.delete();
                        }
                        catch {
                            break;
                        }
                    }
                }
            }
        }

        /// <summary>
        /// 接受请求
        /// </summary>
        /// <param name="param">参数</param>
        private static void readData(object param) {
            Socket socket = (Socket)param;
            FCHttpMonitor nodeService = FCHttpMonitor.MainMonitor;
            int newSocketID = (int)socket.Handle;
            try {
                byte[] buffer = new byte[102400];
                int len = socket.Receive(buffer);
                MemoryStream memoryStream = new MemoryStream(buffer);
                StreamReader reader = new StreamReader(memoryStream);
                FCHttpData data = new FCHttpData();
                data.m_remoteIP = ((IPEndPoint)socket.RemoteEndPoint).Address.ToString();
                data.m_remotePort = ((IPEndPoint)socket.RemoteEndPoint).Port;
                String requestHeader;
                int contentLength = 0;
                string parameters = "";
                while ((requestHeader = reader.ReadLine()) != null && !String.IsNullOrEmpty(requestHeader)) {
                    String lowerHeader = requestHeader.ToLower();
                    if (lowerHeader.IndexOf("get") == 0) {
                        int end = lowerHeader.IndexOf("http/");
                        data.m_method = "GET";
                        parameters = requestHeader.Substring(5, end - 6);
                    }
                    else if (lowerHeader.IndexOf("post") == 0) {
                        int end = lowerHeader.IndexOf("http/");
                        data.m_method = "POST";
                        parameters = requestHeader.Substring(5, end - 6);
                    }
                    else if (lowerHeader.IndexOf("accept: ") == 0) {
                        try {
                            data.m_contentType = requestHeader.Substring(8, requestHeader.IndexOf(',') - 8);
                        }
                        catch { }
                    }
                    else if (lowerHeader.IndexOf("content-type:") == 0) {
                        data.m_contentType = requestHeader.Substring(14);
                    }
                    else if (lowerHeader.IndexOf("host:") == 0) {
                        data.m_url = requestHeader.Substring(requestHeader.IndexOf(':') + 2);
                    }
                    else if (lowerHeader.IndexOf("content-length") == 0) {
                        int begin = lowerHeader.IndexOf("content-length:") + "content-length:".Length;
                        String postParamterLength = requestHeader.Substring(begin).Trim();
                        contentLength = Convert.ToInt32(postParamterLength);
                    }
                }
                if (contentLength > 0) {
                    int idx = 0, ide = 0;
                    data.m_body = new byte[contentLength];
                    while (idx < contentLength) {
                        int recvData = reader.Read();
                        if (recvData != -1) {
                            if (recvData != 0) {
                                ide++;
                            }
                            idx++;
                        }
                        else {
                            break;
                        }
                    }
                    reader.Close();
                    memoryStream.Dispose();
                    if (ide == 0) {
                        socket.Receive(data.m_body);
                    }
                    else {
                        for (int i = 0; i < contentLength; i++) {
                            data.m_body[i] = buffer[len - contentLength + i];
                        }
                    }
                    data.m_contentLength = contentLength;
                }
                else {
                    reader.Close();
                    memoryStream.Dispose();
                }
                if (data.m_method.Length == 0) {
                    return;
                }
                int cindex = parameters.IndexOf('?');
                if (cindex != -1) {
                    data.m_url = data.m_url + "/" + parameters;
                    parameters = parameters.Substring(cindex + 1);
                    String[] strs = parameters.Split(new string[] { "&" }, StringSplitOptions.RemoveEmptyEntries);
                    int strsSize = strs.Length;
                    for (int i = 0; i < strsSize; i++) {
                        String[] subStrs = strs[i].Split(new string[] { "=" }, StringSplitOptions.RemoveEmptyEntries);
                        data.m_parameters.put(subStrs[0].ToLower(), subStrs[1]);
                    }
                }
                else {
                    data.m_url += "/" + parameters;
                }
                FCScript indicator = null;
                if (nodeService.UseScript) {
                    try {
                        lock (nodeService.m_indicators) {
                            indicator = nodeService.m_indicators.Pop();
                        }
                    }
                    catch {
                        indicator = CFunctionEx.CreateScript(nodeService.Script, nodeService.Native);
                    }
                    ArrayList<CFunction> functions = indicator.getFunctions();
                    int functionsSize = functions.Count;
                    for (int i = 0; i < functionsSize; i++) {
                        CFunctionHttp function = functions.get(i) as CFunctionHttp;
                        if (function != null) {
                            function.m_data = data;
                        }
                    }
                }
                data.m_socketID = newSocketID;
                lock (nodeService.m_httpDatas) {
                    nodeService.m_httpDatas.put(newSocketID, data);
                }
                if (indicator != null) {
                    indicator.callFunction("ONHTTPREQUEST();");
                }
                if (data.m_close) {
                    return;
                }
                int resContentLength = 0;
                if (data.m_resBytes != null) {
                    resContentLength = data.m_resBytes.Length;
                }
                else {
                    if (data.m_resStr != null) {
                        resContentLength = Encoding.Default.GetBytes(data.m_resStr).Length;
                    }
                }
                StringBuilder bld = new StringBuilder();
                bld.Append("HTTP/1.0 " + data.m_statusCode.ToString() + " OK\r\n");
                bld.Append(String.Format("Content-Length: {0}\r\n", resContentLength));
                bld.Append("Connection: close\r\n\r\n");
                if (data.m_resBytes != null) {
                    socket.Send(Encoding.Default.GetBytes(bld.ToString()));
                    socket.Send(data.m_resBytes);
                }
                else {
                    bld.Append(data.m_resStr);
                    socket.Send(Encoding.Default.GetBytes(bld.ToString()));
                }
                if (indicator != null) {
                    lock (nodeService.m_indicators) {
                        nodeService.m_indicators.Push(indicator);
                    }
                }
            }
            catch (Exception ex) {
                Console.WriteLine(ex.Message + "\r\n" + ex.StackTrace);
            }
            finally {
                lock (nodeService.m_httpDatas) {
                    nodeService.m_httpDatas.Remove(newSocketID);
                }
                socket.Close();
            }
        }

        /// <summary>
        /// 启动监听
        /// </summary>
        public void start() {
            m_useScript = FCFile.isFileExist(m_fileName);
            if (m_useScript) {
                m_native = new FCNative();
                FCFile.read(m_fileName, ref m_script);
                m_indicator = CFunctionEx.CreateScript(m_script, m_native);
                Console.WriteLine(m_script);
            }
            try {
                //string host = "127.0.0.1";
                //IPAddress ip = IPAddress.Parse(host);
                if (m_indicator != null) {
                    m_indicator.callFunction("ONHTTPSERVERSTARTING('" + m_fileName + "');");
                }
                IPEndPoint ipe = new IPEndPoint(IPAddress.Any, m_port);

                m_listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                m_listener.Bind(ipe);
                m_listener.Listen(0);
            }
            catch (Exception ex) {
                if (m_indicator != null) {
                    m_indicator.callFunction("ONHTTPSERVERSTARTFAIL('" + ex.Message + "\r\n" + ex.StackTrace + "');");
                }
                return;
            }
            int minThreadNum = 0, portThreadNum = 0, maxThreadNum = 0;
            ThreadPool.GetMaxThreads(out maxThreadNum, out portThreadNum);
            ThreadPool.GetMinThreads(out minThreadNum, out portThreadNum);
            if (m_indicator != null) {
                m_indicator.callFunction("ONHTTPSERVERSTART(" + FCStr.convertIntToStr(maxThreadNum) + "," + FCStr.convertIntToStr(minThreadNum) + ");");
            }
            while (true) {
                Socket socket = m_listener.Accept();
                ThreadPool.QueueUserWorkItem(new WaitCallback(readData), socket);
            }
            m_listener.Close();
        }
    }
}

File: facecat_cpp/source/service/FCHttpMonitor.cpp
#include "..\\..\\stdafx.h"
#include "..\\..\\include\\service\\FCHttpMonitor.h"
#include "..\\..\\include\\service\\CFunctionEx.h"
#include "..\\..\\include\\service\\CFunctionHttp.h"
#include "..\\..\\include\\core\\FCBinary.h"

namespace FaceCat{	
	static map<String, FCHttpEasyService*> m_easyServices;


	FCHttpData::FCHttpData(){
		m_body = 0;
		m_close = false;
		m_contentLength = 0;
		m_contentType = L"";
		m_method = L"";
		m_parameters.clear();	
		m_resBytes = 0;
		m_resBytesLength = 0;
		m_resStr = L"";	
		m_remoteIP = L"";
        m_remotePort = 0;
		m_socket = 0;
		m_statusCode = 200;
		m_url = L"";
	}

	FCHttpData::~FCHttpData(){
		m_parameters.clear();
		if(m_body){
			delete m_body;
			m_body = 0;
		}
		if(m_resBytes){
			delete m_resBytes;
			m_resBytes = 0;
		}
	}

	////////////////////////////////////////////////////////////////////////////////////////////////////////////

	FCHttpMonitor::FCHttpMonitor(string fileName){
		m_fileName = fileName;
		m_hSocket = 0;
		m_indicator = 0;
		m_native = 0;
		m_port = 8080;
		m_script = L"";
		m_useScript = true;
	}

	FCHttpMonitor::~FCHttpMonitor(){
		map<int, FCHttpData*>::iterator sIter = m_httpDatas.begin();
		for(;sIter != m_httpDatas.end(); sIter++){
			delete sIter->second;
		}
		m_httpDatas.clear();
		m_indicator = 0;
		while (!m_indicators.empty()){
			FCScript *indicator = m_indicators.top();
			m_indicators.pop();
			delete indicator;
		}  
		m_native = 0;
	}

	FCHttpEasyService* FCHttpMonitor::getEasyService(const String& name){
		if(m_easyServices.find(name) != m_easyServices.end()){
			return m_easyServices[name];
		}
		else{
			return 0;
		}
	}

	FCScript* FCHttpMonitor::getIndicator(){
		return m_indicator;
	}

	static FCHttpMonitor *m_mainMonitor = 0;

	FCHttpMonitor* FCHttpMonitor::getMainMonitor(){
		return m_mainMonitor;
	}

	FCNative* FCHttpMonitor::getNative(){
		return m_native;
	}

	int FCHttpMonitor::getPort(){
		return m_port;
	}

	void FCHttpMonitor::setPort(int port){
		m_port = port;
	}

	SOCKET FCHttpMonitor::getSocket(){
		return m_hSocket;
	}

	String FCHttpMonitor::getScript(){
		return m_script;
	}

	void FCHttpMonitor::setScript(String script)
	{
		m_script = script;
	}

	bool FCHttpMonitor::getUseScript(){
		return m_useScript;
	}

	void FCHttpMonitor::setHttpData(int socketID, FCHttpData* data){
		m_lockHttpData.lock();
		m_httpDatas.insert(make_pair(socketID, data));
		m_lockHttpData.unLock();
	}

	FCScript* FCHttpMonitor::popCIndicator(){		
		FCScript* indicator = 0;
		m_lock.lock();		
		if (!m_indicators.empty()){
			indicator = m_indicators.top();
			m_indicators.pop();
		}
		m_lock.unLock();
		return indicator;
	}

	//////////////////////////////////////////////////////////////////////////////////////////////////////////////

	DWORD WINAPI receiveHandle(LPVOID lpParam){
		FCHttpData *data = (FCHttpData*)lpParam;
		if(!data) return 0;		
		SOCKADDR_IN addr = data->m_addr;	
		const int bufLen = 102400;
		char buffer[bufLen];
		memset(buffer, '\0', bufLen);
        String parameters = L"";
		int contentLength = 0;
		int copyLen = recv(data->m_socket, buffer, bufLen, 0);
		if(copyLen == 0 || copyLen == SOCKET_ERROR) {
			return 0;
		}		
		strstream os(buffer, copyLen);
		while(true){
			char bufferline[bufLen];
			memset(bufferline, '\0', bufLen);
			os.getline(bufferline, bufLen);
			string requestHeader(bufferline);
			int headLen = requestHeader.length();
			if(headLen == 0 || (headLen == 1 && requestHeader[0] == '\r')){
				break;
			}
			int keyLength = 0;
			int begin = 5;
			if(requestHeader.find("GET") == 0){
				int end = (int)requestHeader.find("HTTP/");
				data->m_method = L"GET";
				parameters = FCStr::stringTowstring(requestHeader.substr(begin, end - 6));
			}
			else if(requestHeader.find("POST") == 0){
				int end = (int)requestHeader.find("HTTP/");
				data->m_method = L"POST";
				parameters = FCStr::stringTowstring(requestHeader.substr(begin, end - 6));
			}
			else if(requestHeader.find("Accept:") == 0){
				int acceptIndex = 0;
				keyLength = (int)strlen("Accept:");
				begin = acceptIndex + keyLength;
				data->m_contentType = FCStr::stringTowstring(requestHeader.substr(begin, requestHeader.find(',', begin) - begin));
			}
			else if (requestHeader.find("Host:") == 0){
				int hostIndex = 0;
				keyLength = (int)strlen("Host:");
				begin = hostIndex + keyLength;
				data->m_url = FCStr::stringTowstring(requestHeader.substr(begin, requestHeader.find("\r\n", begin) - begin));
			}
			else if (requestHeader.find("Content-Length:") == 0){
				int contentIndex = 0;
				keyLength = (int)strlen("Content-Length:");
				begin = contentIndex + keyLength;
				string postParamterLength = FCStr::replace(requestHeader.substr(begin), " ", "");;
				contentLength = atoi(postParamterLength.c_str());
			}
		}		
        if (contentLength > 0){
            int idx = 0, ide = 0;
            data->m_body = new char[contentLength];
            while (idx < contentLength){
				int recvData = os.get();
                if (recvData != -1){
                    if (recvData != 0){
                        ide++;
                    }
                    idx++;
                }
                else{
                    break;
                }
            }
            if (ide == 0){
                recv(data->m_socket, data->m_body, contentLength, 0);
            }
            else{
                for (int i = 0; i < contentLength; i++){
                    data->m_body[i] = buffer[copyLen - contentLength + i];
                }
            }
            data->m_contentLength = contentLength;
        }
		if ((int)data->m_method.length() == 0){
			closesocket(data->m_socket);
			delete data;
            return 0;
        }
		int cindex = (int)parameters.find('?');
        if (cindex != -1){
            data->m_url = data->m_url + L"/" + parameters;
			parameters = parameters.substr(cindex + 1);
			ArrayList<String> strParameters = FCStr::split(parameters, L"&");
			int strsSize = (int)strParameters.size();
            for (int i = 0; i < strsSize; i++){
				ArrayList<String> subStrs = FCStr::split(strParameters.get(i), L"=");
				if((int)subStrs.size() > 1){
					string value = FCStr::wstringTostring(subStrs.get(1));
					value = FCStr::urlDecode(value);
					String wValue = FCStr::stringTowstring(value);
					String wPara = subStrs.get(0);
					data->m_parameters.insert(make_pair(wPara, wValue));
				}
            }
        }
        else{
            data->m_url += L"/" + parameters;
        }		
        FCScript *indicator = 0;
		if (m_mainMonitor->getUseScript()){
			indicator = m_mainMonitor->popCIndicator();
			if(indicator == 0){
				indicator = CFunctionEx::createIndicator(m_mainMonitor->getScript(), m_mainMonitor->getNative());
			}
			ArrayList<CFunction*> functions = indicator->getFunctions();
			int functionsSize = (int)functions.size();
            for (int i = 0; i < functionsSize; i++){
				CFunctionHttp *function = dynamic_cast<CFunctionHttp*>(functions.get(i));
                if (function != 0){
					function->setHttpData(data);
                }
            }
        }
		m_mainMonitor->setHttpData((int)data->m_socket, data);
		if(indicator != 0){
			indicator->callFunction(L"ONHTTPREQUEST();");
		}
        if (data->m_close){
			closesocket(data->m_socket);
			delete data;
            return 0;
        }
		int resContentLength = 0;
        if (data->m_resBytes){
            resContentLength = data->m_resBytesLength;
        }
        else{
			string str = FCStr::wstringTostring(data->m_resStr);
            resContentLength = (int)str.length() + 1;
        }
		string retMessage = "";
		char statusCode[10]; 		
		sprintf_s(statusCode, 9, "%d", data->m_statusCode);
		char szContentLength[10]; 		
		sprintf_s(szContentLength, 9, "%d", resContentLength);
		retMessage = retMessage.append("HTTP/1.0 ");
		retMessage = retMessage.append(statusCode);
		retMessage = retMessage.append(" OK\r\n");
		retMessage = retMessage.append("Content-Length:");
		retMessage = retMessage.append(szContentLength);
		retMessage = retMessage.append("\r\n");
		retMessage = retMessage.append("Connection: close\r\n\r\n");
		if (data->m_resBytes){
			int retLength = (int)strlen(retMessage.c_str());
			int ret = send(data->m_socket, retMessage.c_str(), retLength, 0);
			ret = send(data->m_socket, data->m_resBytes, data->m_resBytesLength, 0);
        }
        else{
			string str = FCStr::wstringTostring(data->m_resStr);
			retMessage = retMessage.append(str);
			int ret = send(data->m_socket, retMessage.c_str(), (int)retMessage.length() + 1, 0);
        }
		closesocket(data->m_socket);
		delete data;
		return 0;
	}

	///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
	
	void FCHttpMonitor::checkScript(){
		String wFileName = FCStr::stringTowstring(m_fileName);
        String newScript = L"";
		FCFile::read(wFileName, &newScript);
		if (m_script.compare(newScript) == 0){
			cout << L"检测到脚本被修改...";
            m_script = newScript;			
			m_lock.lock();			
            while (true){
				FCScript *indicator = m_indicators.top();
				m_indicators.pop();
                delete indicator;
            }
			m_lock.unLock();
        }
	}

	int FCHttpMonitor::close(int socketID){
		SOCKET socket = (SOCKET)socketID;
		int ret = 0;
		if(socket){
			ret = closesocket(socket);
		}
		return ret;
	}

	int FCHttpMonitor::start(string fileName){
		String wFileName = FCStr::stringTowstring(m_fileName);
		m_useScript = FCFile::isFileExist(wFileName);
        if (m_useScript){
            m_native = new FCNative;
			FCFile::read(wFileName, &m_script);
			m_indicator = CFunctionEx::createIndicator(m_script, m_native);
			cout << m_script.c_str() << "\r\n";
        }
		if(m_indicator != 0){
			String funcName = L"ONHTTPSERVERSTARTING('";
			funcName = funcName.append(wFileName);
			funcName = funcName.append(L"');");
			m_indicator->callFunction(funcName);
		}		
		int ret = 0;
		m_hSocket = socket(AF_INET, SOCK_STREAM, 0);
		m_addr.sin_addr.S_un.S_addr = INADDR_ANY;
		m_addr.sin_family = AF_INET;
		m_addr.sin_port = htons(m_port);
		ret = bind(m_hSocket, (SOCKADDR*)&m_addr, sizeof(SOCKADDR));
		if(ret != SOCKET_ERROR){
			ret = listen(m_hSocket, 2);
			if(ret != SOCKET_ERROR){
				while(true){
					SOCKADDR_IN addr;
					int addrlen = sizeof(addr);
					SOCKET socket = accept(m_hSocket, (LPSOCKADDR)&addr, &addrlen);
					if(socket == SOCKET_ERROR){
						::Sleep(1);
						continue;
					}
					FCHttpData *data = new FCHttpData;
					data->m_addr = addr;
					data->m_socket = socket;
					HANDLE hThread = ::CreateThread(NULL, 0, receiveHandle, (LPVOID)data, 0, 0);
					::CloseHandle(hThread);
				}
			}
		}
		if(m_hSocket){
			ret = closesocket(m_hSocket);
		}
		return ret;
	}
}